test/integration/storageversion/storage_version_filter_test.go

```diff
@@ -0,0 +1,235 @@
+	assertBlocking("reads of built-in resources", t, err, shouldBlock)
+}
+
+// TestStorageVersionBootstrap ensures that before the StorageVersions are
+// updated, only the following the request are accepted by the apiserver:
+// 1. read requests
+// 2. non-persisting write requests
+// 3. write requests to the storageversion API
+// 4. requests to CR or aggregated API
+func TestStorageVersionBootstrap(t *testing.T) {
+	// Start server and create CRD
+	etcdConfig := framework.SharedEtcd()
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, etcdConfig)
+	etcd.CreateTestCRDs(t, apiextensionsclientset.NewForConfigOrDie(server.ClientConfig), false, etcd.GetCustomResourceDefinitionData()[0])
+	server.TearDownFn()
+
+	startUpdateSV := make(chan struct{})
+	finishUpdateSV := make(chan struct{})
+	delegateUpdateFinished := make(chan struct{})
+	completed := make(chan struct{})
+	wrapperFunc := func(delegate storageversion.Manager) storageversion.Manager {
+		return &wrappedStorageVersionManager{
+			startUpdateSV:          startUpdateSV,
+			finishUpdateSV:         finishUpdateSV,
+			delegateUpdateFinished: delegateUpdateFinished,
+			completed:              completed,
+			Manager:                delegate,
+		}
+	}
+	// Restart api server, enable the storage version API and the feature gate.
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageVersionAPI, true)()
+	server = kubeapiservertesting.StartTestServerOrDie(t,
+		&kubeapiservertesting.TestServerInstanceOptions{
+			SkipPostStartChecks:    true,
+			StorageVersionWrapFunc: wrapperFunc,
+		},
+		[]string{
+			// force enable all resources to ensure that the storage updates can handle cross group resources.
+			// TODO: drop these once we stop allowing them to be served.
+			"--runtime-config=api/all=true,extensions/v1beta1/deployments=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/podsecuritypolicies=true,extensions/v1beta1/networkpolicies=true,internal.apiserver.k8s.io/v1alpha1=true",
+		},
+		etcdConfig)
+	defer server.TearDownFn()
+
+	cfg := rest.CopyConfig(server.ClientConfig)
+
+	// Write to k8s native API object should be blocked.
+	testBuiltinResourceWrite(t, cfg, true)
+	// Write to CRD should be blocked.
+	testCRDWrite(t, cfg, true)
+	// Write to APIService should be blocked.
+	testAPIServiceWrite(t, cfg, true)
+	// Write to CR should pass.
+	testCRWrite(t, cfg, false)
+	// Write to the storage version API should pass.
+	testStorageVersionWrite(t, cfg, false)
+	// Write to non-persisted API should pass.
+	testNonPersistedWrite(t, cfg, false)
+	// Read of k8s native API should pass.
+	testBuiltinResourceRead(t, cfg, false)
+
+	// After the storage versions are updated, even though the
+	// StorageVersionManager.Complete() still returns false, the filter
+	// should not block any request.
+	close(startUpdateSV)
+	<-delegateUpdateFinished
```

Comment:
Can we rename this variable to `updateFinished`? I don't get what "delegate" means here, though probably I came up with this name :)